src/MultiMessageSender.sol

```diff
@@ -133,104 +148,165 @@ contract MultiMessageSender {
+        address[] memory filtered = _filterAdapters(existings, _removals);
+
+        senderAdapters = filtered;
+        _logSenderAdapterUpdates(_removals, false);
     }
 
     /*/////////////////////////////////////////////////////////////////
-                            EXTERNAL VIEW FUNCTIONS
+                            PRIVATE/INTERNAL FUNCTIONS
     ////////////////////////////////////////////////////////////////*/
 
-    /// @notice A helper function for estimating total required message fee by all available message bridges.
-    function estimateTotalMessageFee(
-        uint256 _dstChainId,
-        address _multiMessageReceiver,
-        address _target,
-        bytes calldata _callData,
-        uint256 _nativeValue
-    ) public view returns (uint256 totalFee) {
-        MessageLibrary.Message memory message =
-            MessageLibrary.Message(block.chainid, _dstChainId, _target, nonce, _callData, _nativeValue, 0);
-        bytes memory data;
+    function _remoteCall(RemoteCallArgs memory _args) private {
+        (address mmaReceiver, address[] memory adapters) = _checkRemoteCallArgs(
+            _args.dstChainId, _args.target, _args.refundAddress, _args.fees, _args.excludedAdapters
+        );
 
-        /// @dev writes to memory for saving gas
-        address[] storage adapters = senderAdapters;
+        /// @dev increments nonce
+        ++nonce;
 
-        /// @dev generates the dst chain function call
-        data = abi.encodeWithSelector(IMultiMessageReceiver.receiveMessage.selector, message);
+        MessageLibrary.Message memory message = MessageLibrary.Message(
+            block.chainid,
+            _args.dstChainId,
+            _args.target,
+            nonce,
+            _args.callData,
+            _args.nativeValue,
+            block.timestamp + _args.expiration
+        );
+        bytes32 msgId = MessageLibrary.computeMsgId(message);
+        bool[] memory adapterSuccess = _dispatchMessages(adapters, mmaReceiver, _args.dstChainId, message, _args.fees);
 
-        for (uint256 i; i < adapters.length; ++i) {
-            uint256 fee =
-                IMessageSenderAdapter(adapters[i]).getMessageFee(uint256(_dstChainId), _multiMessageReceiver, data);
+        emit MultiMessageSent(
+            msgId,
+            nonce,
+            _args.dstChainId,
+            _args.target,
+            _args.callData,
+            _args.nativeValue,
+            _args.expiration,
+            adapters,
+            adapterSuccess
+        );
 
-            totalFee += fee;
+        /// refund remaining fee
+        if (address(this).balance > 0) {
+            _safeTransferETH(_args.refundAddress, address(this).balance);
         }
     }
 
-    /*/////////////////////////////////////////////////////////////////
-                            PRIVATE/INTERNAL FUNCTIONS
-    ////////////////////////////////////////////////////////////////*/
-
-    function _remoteCall(
+    function _checkRemoteCallArgs(
```

Comment:
nit: suggest renaming to `_checkAndProcessArgs()`

Reply:
Done.